src/server/frontend_wayland/wl_pointer.cpp

```diff
@@ -112,15 +112,8 @@ void mf::WlPointer::handle_event(MirPointerEvent const* event, WlSurface* surfac
             auto point = Point{mir_pointer_event_axis_value(event, mir_pointer_axis_x),
                                         mir_pointer_event_axis_value(event, mir_pointer_axis_y)};
             auto transformed = surface->transform_point(point);
-            if (transformed)
-            {
-                handle_enter(transformed.value().first, transformed.value().second);
-                handle_frame();
-            }
-            else
-            {
-                log_warning("surface->transform_point() returned nullopt for enter action");
-            }
+            handle_enter(transformed.position, transformed.surface);
```

Comment:
In general, this function should only be called with input inside the input region, as Mir is already using that to determine which surface to dispatch input to. If there is some case where it's getting input outside the region, clients should be able to handle it.

Reply:
Is `input_shape` communicated up to the rest of Mir? If not, this could receive an event that's inside the surface but not inside the input region and incorrectly send an enter event for it?

Reply:
The input shape is communicated to the Mir scene, and seems to be working "well enough"[1]. All the Wayland code needs to do is identify the correct subsurface to deliver the event to and transform the co-ordinates accordingly.

[1] I don't remember seeing any tests ensuring correct working, so there are likely some corner cases that need fixing (like this).